src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,338 @@
+            {
+                if (predicate(arguments[i]))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(
+            IParameterSymbol lastParameter,
+            ImmutableArray<IArgumentOperation> arguments,
+            INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.Type.Equals(cancellationTokenType) &&
+                lastParameter.IsOptional) // Has a default value being used
+            {
+                // Find out if the ct argument is using the default value
+                // Need to check among all arguments in case the user is passing them named and unordered (despite the ct being defined as the last parameter)
+                return AnyArgument(
+                    arguments,
+                    a => a.Parameter.Equals(cancellationTokenType) && a.ArgumentKind == ArgumentKind.DefaultValue);
+            }
+            return false;
+        }
+
+        // Checks if the method has a `params CancellationToken[]` argument in the last position and ensure no ct is being passed.
+        private static bool InvocationIsUsingParamsCancellationToken(
+            IParameterSymbol lastParameter,
+            ImmutableArray<IArgumentOperation> arguments,
+            INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.IsParams &&
+                   lastParameter.Type is IArrayTypeSymbol arrayTypeSymbol &&
+                   arrayTypeSymbol.ElementType.Equals(cancellationTokenType))
+            {
+                IArgumentOperation? paramsArgument = arguments.FirstOrDefault(a => a.ArgumentKind == ArgumentKind.ParamArray);
+                if (paramsArgument?.Value is IArrayCreationOperation arrayOperation)
+                {
+                    // Do not offer a diagnostic if the user already passed a ct to the params
+                    return arrayOperation.Initializer.ElementValues.IsEmpty;
+                }
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(
+            IMethodSymbol method,
+            ITypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out IMethodSymbol? overload)
+        {
+            ImmutableArray<ISymbol> overloads = method.ContainingType.GetMembers(method.Name);
+            System.Collections.Generic.List<IMethodSymbol> ofType = overloads.OfType<IMethodSymbol>().ToList();
```

Comment:
Consider using a pooled array builder:
```suggestion
            using var ofType = ArrayBuilder<IMethodSymbol>.GetInstance();
            ofType.AddRange(overloads.OfType<IMethodSymbol>());
```

Reply:
Actually I should remove these two lines, I accidentally left them like that because I was debugging, but the original code had them in one single call.